src/main/java/org/elasticsearch/transport/TransportService.java

```diff
@@ -609,4 +685,89 @@ public void cancelTimeout() {
             }
         }
     }
+
+    static class DirectResponseChannel implements TransportChannel {
+        final private String action;
+        final private long requestId;
+        final TransportServiceAdapter adapter;
+        final ThreadPool threadPool;
+
+        public DirectResponseChannel(String action, long requestId, TransportServiceAdapter adapter, ThreadPool threadPool) {
+            this.action = action;
+            this.requestId = requestId;
+            this.adapter = adapter;
+            this.threadPool = threadPool;
+        }
+
+        @Override
+        public String action() {
+            return action;
+        }
+
+        @Override
+        public void sendResponse(TransportResponse response) throws IOException {
+            sendResponse(response, TransportResponseOptions.EMPTY);
+        }
+
+        @Override
+        public void sendResponse(final TransportResponse response, TransportResponseOptions options) throws IOException {
+            final TransportResponseHandler handler = adapter.onResponseReceived(requestId);
+            // ignore if its null, the adapter logs it
+            if (handler != null) {
+                final String executor = handler.executor();
+                if (ThreadPool.Names.SAME.equals(executor)) {
```

Comment:
can we pass to the channel the executor the `TransportRequestHandler` is executing on, and on top of SAME, if its the same as the executor the request is executing on, we don't need to fork it again?